src/jaxsim/simulation/integrators.py

```diff
@@ -303,6 +305,172 @@ def body_fun(carry: Carry, xs: None) -> Tuple[Carry, None]:
+) -> Tuple[ODEState, Dict[str, Any]]:
+    """
+    Semi-implicit Euler integrator with quaternion integration on SO(3).
+
+    Args:
+        dx_dt: Callable that computes the state derivative.
+        x0: Initial state as ODEState object.
+        t0: Initial time.
+        tf: Final time.
+        num_sub_steps: Number of sub-steps to break the integration into.
+
+    Returns:
+        A tuple having as first element the final state as ODEState object,
+        and as second element a dictionary including auxiliary data at t0.
+    """
+
+    # Compute the sub-step size.
+    # We break dt in configurable sub-steps.
+    dt = tf - t0
+    sub_step_dt = dt / num_sub_steps
+
+    # Integrate the quaternion on its manifold using the new angular velocity
+
+    # Initialize the carry
+    Carry = Tuple[ODEState, Time]
+    carry_init: Carry = (x0, t0)
+
+    def body_fun(carry: Carry, xs: None) -> Tuple[Carry, None]:
+        # Unpack the carry
+        x_t0, t0 = carry
+
+        # Compute the state derivative.
+        # We only keep the quantities related to the acceleration and discard those
+        # related to the velocity since we are going to use those implicitly integrated
+        # from the accelerations.
+        StateDerivative = ODEState
+        dxdt_t0: StateDerivative = dx_dt(x_t0, t0)[0]
+
+        # Extract the initial position ∈ ℝ⁷⁺ⁿ and initial velocity ∈ ℝ⁶⁺ⁿ.
+        # This integrator, contrarily to most of the other ones, is not generic.
+        # It expects to operate on an x object of class ODEState.
+        pos_t0 = x_t0.physics_model.position()
+        vel_t0 = x_t0.physics_model.velocity()
+
+        # Extract the velocity derivative
+        d_vel_dt = dxdt_t0.physics_model.velocity()
+
+        # =============================================
+        # Perform semi-implicit Euler integration [1-4]
+        # =============================================
+
+        # 1. Integrate the accelerations obtaining the implicit velocities
+        # 2. Compute the derivative of the generalized position (w/o quaternion)
+        # 3. Integrate the implicit velocities (w/o quaternion)
+        # 4. Integrate the remaining state
+        # 5. Outside the loop: integrate the quaternion on SO(3) manifold
+
+        # ----------------------------------------------------------------
+        # 1. Integrate the accelerations obtaining the implicit velocities
+        # ----------------------------------------------------------------
+
+        vel_tf = vel_t0 + sub_step_dt * d_vel_dt
+
+        # ----------------------------------------------------------------------
+        # 2. Compute the derivative of the generalized position (w/o quaternion)
+        # ----------------------------------------------------------------------
+
+        # Compute the transform of the mixed base frame at t0
+        W_H_BW = jnp.vstack(
+            [
+                jnp.block([jnp.eye(3), jnp.vstack(x_t0.physics_model.base_position)]),
```

Comment:
```suggestion
                jnp.block([jnp.eye(3), x_t0.physics_model.base_position]),
```
Is `jnp.vstack` necessary? `base_position` should be `np.array` as a member of [`RootPose`](https://github.com/ami-iit/jaxsim/blob/4f56d8687fb44fee41ab5b56174ae812a71c023d/src/jaxsim/parsers/kinematic_graph.py#L34), so probably a conversion is not needed

Reply:
This is actually [`PhysicsModelState.base_position`](https://github.com/ami-iit/jaxsim/blob/4f56d8687fb44fee41ab5b56174ae812a71c023d/src/jaxsim/physics/model/physics_model_state.py#L31-L33) that needs to become a vertical vector.

Reply:
And I mean, the following code fails:

```python
import jax.numpy as jnp

jnp.block([jnp.eye(3), jnp.zeros(3)])
```

You need to vstack the 3D vector.

Reply:
Got it, thanks